### PR TITLE
Add GPT-NeoX model

### PR DESCRIPTION
**Pull request by nicolov**
Implement the GPT-NeoX model from https://huggingface.co/docs/transformers/en/model_doc/gpt_neox

On M2 Ultra (non quantized):

```
==========
Prompt: what
 is the best way to do this?

A:

You can use the following code:
$('#my_div').html(function(i, html) {
    return html.replace(/<br\s*\/
==========
Prompt: 0.943 tokens-per-sec
Generation: 15.146 tokens-per-sec
```

Closes https://github.com/ml-explore/mlx-examples/issues/847

### CODE REVIEW

File: llms/mlx_lm/models/gpt_neox.py [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+                    f"h.{i}.attn.output.weight"
+                ].transpose(1, 0)
+            if f"h.{i}.mlp.c_fc.weight" in weights:
+                weights[f"h.{i}.mlp.c_fc.weight"] = weights[
+                    f"h.{i}.mlp.c_fc.weight"
+                ].transpose(1, 0)
+            if f"h.{i}.mlp.output.weight" in weights:
+                weights[f"h.{i}.mlp.output.weight"] = weights[
+                    f"h.{i}.mlp.output.weight"
+                ].transpose(1, 0)
```
Comment (angeloskath) [inline llms/mlx_lm/models/gpt_neox.py:209]: I am not sure these are being used, given that there aren't part of the model. Did I miss something?

**Review by angeloskath — COMMENTED** [review]
Thanks! It seems to work great!

I left two comments, let me know wdyt.

**Comment by awni** [comment]
Thanks for the port @nicolov ! Would also be good to add the model type (and possibly default key names) to the [Lora training utility](https://github.com/ml-explore/mlx-examples/blob/main/llms/mlx_lm/tuner/utils.py#L89).

File: llms/mlx_lm/models/gpt_neox.py [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+        B, L, D = x.shape
+
+        qkv = self.query_key_value(x)
+
+        new_qkv_shape = qkv.shape[:-1] + (self.num_attention_heads, 3 * self.head_dim)
+        qkv = qkv.reshape(*new_qkv_shape)
+
+        queries = qkv[..., : self.head_dim].transpose(0, 2, 1, 3)
+        keys = qkv[..., self.head_dim : 2 * self.head_dim].transpose(0, 2, 1, 3)
+        values = qkv[..., 2 * self.head_dim :].transpose(0, 2, 1, 3)
```
[the remark this answers is not in the document]
Comment (nicolov) [reply llms/mlx_lm/models/gpt_neox.py]: Makes sense, thanks.

File: llms/mlx_lm/models/gpt_neox.py (hunk 1 above)
Comment (nicolov) [reply llms/mlx_lm/models/gpt_neox.py]: You're right, thanks.

File: llms/mlx_lm/models/gpt_neox.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (nicolov) [reply llms/mlx_lm/models/gpt_neox.py]: (mostly just copy-pasted from another model)

**Comment by nicolov** [comment]
> Thanks for the port @nicolov ! Would also be good to add the model type (and possibly default key names) to the [Lora training utility](https://github.com/ml-explore/mlx-examples/blob/main/llms/mlx_lm/tuner/utils.py#L89).

Done and added a smoke test for LoRA.

**Review by angeloskath — APPROVED** [review]
🚀

**Pull request merged**